### ISSUE TITLE
Not able to generate meaningful results on gsm8k_synthetic_cot

### ISSUE THREAD

**Issue by yongzx** [casper-hansen/OpenCoconut#4]
Thank you for creating this repo! 

May I know what should be the correct configuration for training your model/setup (since you are using an entirely different model and eval dataset than the original paper)?

I tried your default configuration as well as increasing the training steps 10x to `max_steps=10000`, but the `output_stage_3/checkpoint-10000` has an **accuracy score of 0.006** on your gsm8k_synthetic_cot. Many of the output predictions are `None` such as the following. It also got the answer for your infer.py example wrong.

```
Input: 
It was Trevor's job to collect fresh eggs from the family's 4 chickens every morning.   He got 4 eggs from Gertrude and 3 eggs from Blanche.  Nancy laid 2 eggs as did Martha.  On the way, he dropped 2 eggs.  How many eggs did Trevor have left?
Target: 9
Predicted: None
```

**Comment by casper-hansen** [comment]
Hi @yongzx, I am aware of this issue. I simply implemented Coconut as it was described in the paper and shared this as an early version, hoping others would benefit. There may be some bugs that are not yet discovered in the modeling code

**Comment by casper-hansen** [comment]
In particular, this parts right here is what I am most unsure about and is something I worked a lot on modifying.

1. The slicing of the last hidden state. Why not feed in the full hidden state?
https://github.com/casper-hansen/OpenCoconut/blob/5519bf433e336cea6c0ee0c954fb1a7f06e92797/opencoconut/models/qwen2.py#L45

2. Handling caching/masking in with batch size > 1 (I think this could be buggy)

- https://github.com/casper-hansen/OpenCoconut/blob/5519bf433e336cea6c0ee0c954fb1a7f06e92797/opencoconut/models/qwen2.py#L46-L57
- https://github.com/casper-hansen/OpenCoconut/blob/5519bf433e336cea6c0ee0c954fb1a7f06e92797/opencoconut/models/qwen2.py#L257-L290

**Comment by yongzx** [comment]
Gotcha! thank you for being transparent – I will also look into this

**Comment by thelongestusernameofall** [comment]
> In particular, this parts right here is what I am most unsure about and is something I worked a lot on modifying.
> 
> 1. The slicing of the last hidden state. Why not feed in the full hidden state?
>    https://github.com/casper-hansen/OpenCoconut/blob/5519bf433e336cea6c0ee0c954fb1a7f06e92797/opencoconut/models/qwen2.py#L45
> 2. Handling caching/masking in with batch size > 1 (I think this could be buggy)
> 
> * https://github.com/casper-hansen/OpenCoconut/blob/5519bf433e336cea6c0ee0c954fb1a7f06e92797/opencoconut/models/qwen2.py#L46-L57
> * https://github.com/casper-hansen/OpenCoconut/blob/5519bf433e336cea6c0ee0c954fb1a7f06e92797/opencoconut/models/qwen2.py#L257-L290

the same confusion。
Maybe we can modify and try.
